src/tools/compiletest/src/runtest.rs

```diff
@@ -2096,6 +2099,89 @@ actual:\n\
         }
         fs::remove_dir(path)
     }
+
+    fn run_ui_test(&self) {
+        println!("ui: {}", self.testpaths.file.display());
+
+        let proc_res = self.compile_test();
+
+        let expected_stderr_path = self.expected_output_path("stderr");
+        let expected_stderr = self.load_expected_output(&expected_stderr_path);
+
+        let expected_stdout_path = self.expected_output_path("stdout");
+        let expected_stdout = self.load_expected_output(&expected_stdout_path);
+
+        let normalized_stdout = self.normalize_output(&proc_res.stdout);
+        let normalized_stderr = self.normalize_output(&proc_res.stderr);
+
+        let mut errors = 0;
+        errors += self.compare_output("stdout", normalized_stdout.as_bytes(), &expected_stdout);
+        errors += self.compare_output("stderr", normalized_stderr.as_bytes(), &expected_stderr);
+
+        if errors > 0 {
+            println!("To update references, run this command from build directory:");
+            let relative_path_to_file =
+                self.testpaths.relative_dir
+                              .join(self.testpaths.file.file_name().unwrap());
+            println!("{}/update-references.sh '{}' '{}'",
+                     self.config.src_base.display(),
+                     self.config.build_base.display(),
+                     relative_path_to_file.display());
+            self.fatal(&format!("{} errors occurred comparing output.", errors));
+        }
+    }
+
+    fn normalize_output(&self, output: &str) -> String {
+        let parent_dir = self.testpaths.file.parent().unwrap();
+        let parent_dir_str = parent_dir.display().to_string();
+        output.replace(&parent_dir_str, "$DIR")
+              .replace("\\", "/") // windows, you know.
+    }
+
+    fn expected_output_path(&self, kind: &str) -> PathBuf {
+        let extension = match self.revision {
+            Some(r) => format!("{}.{}", r, kind),
+            None => kind.to_string(),
+        };
+        self.testpaths.file.with_extension(extension)
+    }
+
+    fn load_expected_output(&self, path: &Path) -> Vec<u8> {
+        if !path.exists() {
+            return vec![];
+        }
+
+        let mut result = Vec::new();
+        match File::open(path).and_then(|mut f| f.read_to_end(&mut result)) {
+            Ok(_) => result,
+            Err(e) => {
+                self.fatal(&format!("failed to load expected output from `{}`: {}", path.display(), e))
+            }
+        }
+    }
+
+    fn compare_output(&self, kind: &str, actual: &[u8], expected: &[u8]) -> usize {
+        if self.config.verbose {
+            println!("normalized {}:\n{}\n", kind, str::from_utf8(actual).unwrap_or("not utf8"));
+            println!("expected {}:\n{}\n", kind, str::from_utf8(expected).unwrap_or("not utf8"));
```

Comment:
May be worth using `String::from_utf8_lossy` instead of "not uf8" on invalid utf-8